csharp/src/Google.Protobuf/ExtensionSet.cs

```diff
@@ -0,0 +1,294 @@
+        {
+            if (set == null)
+            {
+                return false;
+            }
+            IExtensionValue extensionValue;
+            if (set.ValuesByNumber.TryGetValue(WireFormat.GetTagFieldNumber(stream.LastTag), out extensionValue))
+            {
+                extensionValue.MergeFrom(stream);
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Merges the second set into the first set, creating a new instance if first is null
+        /// </summary>
+        public static void MergeFrom<TTarget>(ref ExtensionSet<TTarget> first, ExtensionSet<TTarget> second) where TTarget : IExtensionMessage<TTarget>
+        {
+            if (second == null)
+            {
+                return;
+            }
+            if (first == null)
+            {
+                first = new ExtensionSet<TTarget>();
+            }
+            foreach (var pair in second.ValuesByIdentifier)
+            {
+                IExtensionValue value;
+                if (first.ValuesByIdentifier.TryGetValue(pair.Key, out value))
+                {
+                    value.MergeFrom(pair.Value);
+                }
+                else
+                {
+                    var cloned = pair.Value.Clone();
+                    first.ValuesByIdentifier[pair.Key] = cloned;
+                    first.ValuesByNumber[pair.Key.FieldNumber] = cloned;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Clones the set into a new set. If the set is null, this returns null
+        /// </summary>
+        public static ExtensionSet<TTarget> Clone<TTarget>(ExtensionSet<TTarget> set) where TTarget : IExtensionMessage<TTarget>
+        {
+            if (set == null)
+            {
+                return null;
+            }
+
+            var newSet = new ExtensionSet<TTarget>();
+            foreach (var pair in set.ValuesByIdentifier)
+            {
+                var cloned = pair.Value.Clone();
+                newSet.ValuesByIdentifier[pair.Key] = cloned;
+                newSet.ValuesByNumber[pair.Key.FieldNumber] = cloned;
+            }
+            return newSet;
+        }
+    }
+
+    /// <summary>
+    /// Used for keeping track of extensions in messages. 
+    /// <see cref="IExtensionMessage{T}"/> methods route to this set.
+    /// 
+    /// Most users will not need to use this class directly
+    /// </summary>
+    /// <typeparam name="TTarget">The message type that extensions in this set target</typeparam>
+    public sealed class ExtensionSet<TTarget> where TTarget : IExtensionMessage<TTarget>
+    {
+        internal Dictionary<Extension, IExtensionValue> ValuesByIdentifier { get; } = new Dictionary<Extension, IExtensionValue>();
+        internal Dictionary<int, IExtensionValue> ValuesByNumber { get; } = new Dictionary<int, IExtensionValue>();
```

Comment:
I think only one Dictionary is needed?

Reply:
I used two since one is keyed to each value by reference (for accessing values by identifier) and the other is keyed to each value by tag (for accessing values by input stream tag). Thus instead of having to manually search one dictionary for each value by reference or tag, it's fast for both. The trade-off being slightly more memory usage.

Reply:
It is easy to get the field number from reference: extension.FieldNumber. Only keep ValuesByNumber will be good enough?

Reply:
Ah, I see what you're saying. You're right, one dictionary will work.